DRAFT: GRIM: Support playing with mouse - DO NOT MERGE

orgads opened the pull request:
Originally done by Tobias Pfaff (@tobiaspfaff).

Source repo: https://github.com/tobiaspfaff/grim_mouse

Blogpost: http://tobiaspfaff.com/blog/2014/05/25/grim/

I kept it up-to-date, and cleaned up. Tested on Windows only.

orgads commented:
The hotspots generator scripts/data can be moved elsewhere (scummvm-tools? Somewhere else?)

I'm unsure about all the Android icons and if/why they're needed.

orgads commented:
I tried reaching out for Tobias asking for details about these changes 5 months ago, but he never replied.

aquadran commented:
That patches were never merged because it was hack. Things are not changed on this matter. Unless changes in this PR were reworked.

orgads commented:
Was it ever proposed? I can't find any PR here or on https://github.com/residualvm/residualvm.

I'd like to improve it. Do you have specific comments?

aquadran commented:
It was never propose on PR. But we saw that patches in past. I general things needs fit into engine smoothly and not hack way. You need to review code and see what things do and how they works. Blinding accepting code without work on it is not an option. There must be someone who are willing to work on it and fix it/ improve things. From what I remember this functionally was not completed or fully worked.

BLooperZ commented:
Thanks @orgads

> From what I remember this functionally was not completed or fully worked. 
@aquadran It actually works quite well from my experience (I have played with @orgads updated and cleaned up branch before)

I also think the generator should be moved to devtools.
(Does it includes the build process of mouse.lab?)

probably should also add fallback to old behavior (= no mouse support) when mouse.lab is not available.

orgads commented:
@BLooperZ mouse.lab is embedded in scummvm executable.

@aquadran I cleaned up a lot of debugging/hacks/datafiles, and reverted Android changes completely. The patch is much smaller now.

I consider the patch ready for initial review.

orgads commented on engines/grim/actor.cpp:
```diff
@@ -660,7 +660,7 @@ void Actor::walkTo(const Math::Vector3d &p) {
 					break;
 				}
-
+				//warning("cs %s",sector->getName().c_str());
```
Revert

orgads commented on engines/grim/lua_v1_graphics.cpp:
```diff
@@ -263,6 +268,10 @@ void Lua_V1::ChangePrimitive() {
 		return;
 
 	PrimitiveObject *pmodify = getprimitive(param1);
+	if (!pmodify) {
+		warning("empty primitive !");
+		return;
+	}
```
Revert

aquadran commented on graphics/projection.cpp:
```diff
@@ -0,0 +1,73 @@
+
+namespace Graphics {
+
+bool project(const Math::Vector3d &objPos,
```
this is duplication for math code

orgads replied:
Removed

aquadran commented on graphics/projection.cpp:
```diff
@@ -0,0 +1,73 @@
+	return true;
+}
+
+bool unProject(const Math::Vector3d &winPos,
```
this is duplication for math code

orgads replied:
Removed

aquadran commented on graphics/module.mk:
```diff
@@ -82,6 +82,12 @@ MODULE_OBJS += \
 	tinygl/ztriangle.o \
 	tinygl/zblit.o \
 	tinygl/zdirtyrect.o
+#   tinygl/ztriangle_shadow.o
```
this should not be there

orgads replied:
Done

aquadran commented on engines/grim/sector.cpp:
```diff
@@ -375,11 +375,10 @@ Common::List<Math::Line3d> Sector::getBridgesTo(Sector *sector) const {
 			Math::Vector3d cross_b2 = Math::Vector3d::crossProduct(edge, delta_b2);
-			bool b1_out = cross_b1.dotProduct(_normal) < 0;
-			bool b2_out = cross_b2.dotProduct(_normal) < 0;
+			bool b1_out = cross_b1.dotProduct(_normal) < -1e-7;
```
why is changed?

orgads replied:
https://github.com/tobiaspfaff/grim_mouse/issues/9

aquadran replied:
I'll check that later

aquadran replied:
can you make separated PR for this?

aquadran commented on engines/grim/resource.cpp:
```diff
@@ -123,8 +126,11 @@ ResourceLoader::ResourceLoader() {
 		} else {
 			if (!SearchMan.hasFile("grim-patch.lab"))
 				error("%s", "grim-patch.lab not found");
+			if (!SearchMan.hasFile("mouse.lab"))
```
this collide with remastered

orgads replied:
Done

aquadran commented on engines/grim/lua_v1_mouse.cpp:
```diff
@@ -0,0 +1,121 @@
+	int mode = lua_getnumber(lua_getparam(1));
+	g_grim->getHotspotMan()->switchMode(mode);
+	if (mode == 1 || mode == 5) {
+		if (!lua_isnumber(lua_getparam(2))) return;
```
formatting

orgads replied:
Done

aquadran commented on engines/grim/lua_v1_mouse.cpp:
```diff
@@ -0,0 +1,121 @@
+		int height = lua_getnumber(lua_getparam(6));
+		g_grim->getHotspotMan()->setupDialog(x0, y0, width, height, lines, 1);
+	} else if (mode == 3) {
+		if (!lua_isnumber(lua_getparam(2))) return;
```
formatting

orgads replied:
Done

aquadran commented on engines/grim/lua_v1_graphics.cpp:
```diff
@@ -509,6 +520,7 @@ void Lua_V1::SetGamma() {
 }
 
 void Lua_V1::Display() {
+	g_grim->drawCursor();
```
this collide with remastered

orgads replied:
Last time I asked, I was told that remastered is not supported.

aquadran replied:
yes, but it should not be executed for remaster

orgads replied:
I added an override in `Lua_Remastered`.

aquadran replied:
add check for remastered, do not split

aquadran commented on math/matrix4.cpp:
```diff
@@ -129,6 +129,134 @@ void Matrix<4, 4>::invertAffineOrthonormal() {
 	setPosition(position);
 }
 
+/*
```
from where it was taken? license?

orgads replied:
From a bugreport in Mesa. I suppose the license is public domain.

Anyway, it is not used at all. Removed.

aquadran commented on engines/grim/lua_v1_set.cpp:
```diff
@@ -252,6 +254,7 @@ void Lua_V1::MakeCurrentSet() {
 	const char *name = lua_getstring(nameObj);
 	Debug::debug(Debug::Engine, "Entered new scene '%s'.", name);
 	g_grim->setSet(name);
+	g_grim->getHotspotMan()->updatePerspective();
```
this collide with remastered

orgads replied:
Done

aquadran commented on engines/grim/lua_v1_graphics.cpp:
```diff
@@ -465,10 +470,16 @@ void Lua_V1::KillPrimitive() {
 }
 
 void Lua_V1::DimScreen() {
+	lua_Object level = lua_getparam(1);
```
why this is added?

orgads replied:
No idea. It is part of a large commit https://github.com/tobiaspfaff/grim_mouse/commit/5ef5d6789c14be37dd95eaa5e5536097fc96fffa

aquadran replied:
then, please revert. it's for dimming screen

aquadran commented on engines/grim/lua_v1_graphics.cpp:
```diff
@@ -465,10 +470,16 @@ void Lua_V1::KillPrimitive() {
 
 	PrimitiveObject *prim = getprimitive(primObj);
-	delete prim;
+	if (prim)
```
this is not needed

orgads replied:
Reverted

aquadran commented on engines/grim/lua_v1.cpp:
```diff
@@ -998,7 +1007,7 @@ struct luaL_reg grimPrimitivesOpcodes[] = {
 	{ "ChangePrimitive", LUA_OPCODE(Lua_V1, ChangePrimitive) },
 	{ "KillPrimitive", LUA_OPCODE(Lua_V1, KillPrimitive) },
-	{ "PurgePrimitiveQueue", LUA_OPCODE(Lua_V1, PurgePrimitiveQueue) }
+	{ "PurgePrimitiveQueue", LUA_OPCODE(Lua_V1, PurgePrimitiveQueue) },
```
this line is not needed

orgads replied:
Reverted

aquadran commented on engines/grim/imuse/imuse_track.cpp:
```diff
@@ -8,12 +8,12 @@
  * as published by the Free Software Foundation; either version 2
  * of the License, or (at your option) any later version.
- *
+
```
???

orgads replied:
Done

aquadran commented on engines/grim/imuse/imuse_track.cpp:
```diff
@@ -104,7 +104,7 @@ bool Imuse::startSound(const char *soundName, int volGroupId, int hookId, int vo
 	for (i = 0; i < MAX_IMUSE_TRACKS; i++) {
 		// Filenames are case insensitive, see findTrack
-		if (!scumm_stricmp(_track[i]->soundName, soundName)) {
+		if (!scumm_stricmp(_track[i]->soundName, soundName) && !_track[i]->toBeRemoved) {
```
????

orgads replied:
Reverted

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,752 @@
+			int gid = hs._objId;
+			if ((_ctrlMode != Normal && _ctrlMode != NoWalk) ||
+				hs._type >= 10 || hs._type == 3 || button == 0) continue;
+			if (gid<=0 || !_hotobject[gid]._active) continue;
```
formatting

orgads replied:
Done

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -389,15 +396,23 @@ Common::Error GrimEngine::run() {
 
 	if (getGameType() == GType_GRIM && (g_grim->getGameFlags() & ADGF_REMASTERED)) {
 		g_driver = createRenderer(1600, 900);
+	} else if (ConfMan.hasKey("gameWidth") && ConfMan.hasKey("gameHeight")) {
```
please remove it, it should not be there at all

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -2004,8 +2017,8 @@ void GfxOpenGLS::prepareMovieFrame(Graphics::Surface* frame) {
 	glBindTexture(GL_TEXTURE_2D, _smushTexId);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
this is part of movie playback, please revert change

orgads replied:
Do you want to revert https://github.com/tobiaspfaff/grim_mouse/commit/8a6806afb71c182f4f2d28f8e9e9a923b5704a00 entirely?

aquadran replied:
yep

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -1919,25 +1932,25 @@ void GfxOpenGLS::drawRectangle(const PrimitiveObject *primitive) {
-	float y3 = primitive->getP3().y * _scaleH;
-	float x4 = primitive->getP4().x * _scaleW;
-	float y4 = primitive->getP4().y * _scaleH;
+	float x1 = primitive->getP1().x;// * _scaleW;
```
it's not related to mouse support, please revet

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -1919,25 +1932,25 @@ void GfxOpenGLS::drawRectangle(const PrimitiveObject *primitive) {
-	float y1 = primitive->getP1().y * _scaleH;
-	float x2 = primitive->getP2().x * _scaleW;
-	float y2 = primitive->getP2().y * _scaleH;
+	float x1 = primitive->getP1().x;// * _scaleW;
```
it's not related to mouse support, please revet

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -1897,10 +1910,10 @@ void GfxOpenGLS::drawGenericPrimitive(const float *vertices, uint32 numVertices,
-	float y1 = primitive->getP1().y * _scaleH;
-	float x2 = primitive->getP2().x * _scaleW;
-	float y2 = primitive->getP2().y * _scaleH;
+	float x1 = primitive->getP1().x; //* _scaleW;
```
it's not related to mouse support, please revet

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -1565,8 +1578,8 @@ void GfxOpenGLS::createFont(Font *font) {
 	glBindTexture(GL_TEXTURE_2D, texture[0]);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
it's not related to mouse support, please revet

aquadran commented on engines/grim/gfx_opengl_shaders.cpp:
```diff
@@ -1436,6 +1441,14 @@ void GfxOpenGLS::drawBitmap(const Bitmap *bitmap, int dx, int dy, uint32 layer)
 		shader->setUniform("offsetXY", Math::Vector2d(float(dx) / _gameWidth, float(dy) / _gameHeight));
 		shader->setUniform("sizeWH", Math::Vector2d(width / _gameWidth, height / _gameHeight));
 		shader->setUniform("texcrop", Math::Vector2d(width / nextHigher2((int)width), height / nextHigher2((int)height)));
+		if (bitmap->_data->_canRotate) {
```
I don't remember right now, is mouse cursor bitmap rotated?

aquadran commented on engines/grim/gfx_opengl.cpp:
```diff
@@ -2288,6 +2322,47 @@ void GfxOpenGL::setBlendMode(bool additive) {
 	}
 }
 
+bool GfxOpenGL::worldToScreen(const Math::Vector3d &vec, int &x, int &y) {
```
please port to tinygl. generally just add 't' prefix

aquadran commented on engines/grim/gfx_opengl.cpp:
```diff
@@ -2113,6 +2117,36 @@ void GfxOpenGL::irisAroundRegion(int x1, int y1, int x2, int y2) {
 	glDepthMask(GL_TRUE);
 }
 
+void GfxOpenGL::blackbox(int x0, int y0, int x1, int y1, float opacity) {
```
please port to tinygl. generally just add 't' prefix

aquadran commented on engines/grim/gfx_opengl.cpp:
```diff
@@ -1789,10 +1793,10 @@ void GfxOpenGL::prepareMovieFrame(Graphics::Surface *frame) {
-		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
-		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
this is part of smush, please revert

aquadran commented on engines/grim/gfx_opengl.cpp:
```diff
@@ -1403,10 +1407,10 @@ void GfxOpenGL::createFont(Font *font) {
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
this is part of font, please revert

aquadran commented on engines/grim/gfx_opengl.cpp:
```diff
@@ -46,7 +46,6 @@
 #include "engines/grim/remastered/overlay.h"
 #include "engines/grim/registry.h"
 
-
```
unnecessary change

aquadran commented on engines/grim/gfx_base.h:
```diff
@@ -96,6 +96,8 @@ class GfxBase {
 
 	virtual uint getScreenWidth() { return _screenWidth; }
 	virtual uint getScreenHeight() { return _screenHeight; }
+	float getScaleW() const { return _scaleW; }
```
this is no longer valid, there is no smaller any more

aquadran commented on engines/grim/costume/chore.cpp:
```diff
@@ -196,6 +196,12 @@ void Chore::fade(Animation::FadeMode mode, uint msecs) {
 }
 
 void Chore::fadeIn(uint msecs) {
+	if (!_playing) {
```
why?

aquadran replied:
can you check without it how it works?

aquadran commented on engines/grim/actor.cpp:
```diff
@@ -751,6 +751,10 @@ void Actor::walkTo(const Math::Vector3d &p) {
 					_walking = false;
 					return;
 				}
+				if (!forceWalk && getName() == "Manny") {
```
add check for remastered

aquadran commented on engines/grim/remastered/lua_remastered.cpp:
```diff
@@ -565,6 +567,25 @@ void Lua_Remastered::Save() {
+	}
+}
+
+void Lua_Remastered::MakeCurrentSet() {
```
don't split, just add check for remastered

aquadran commented on engines/grim/lua_v1_graphics.cpp:
```diff
@@ -38,13 +38,18 @@ namespace Grim {
 
 void Lua_V1::GetImage() {
 	lua_Object nameObj = lua_getparam(1);
+	lua_Object transObj = lua_getparam(2);
```
I guess second param is not present in original, code should check for presence

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,768 @@
+void HotspotMan::renameHotspot(int num, const Common::String &name) {
+	Common::String id = name, desc = name;
+	for (size_t i = 1, total = id.size(); i < total; ++i) {
+		if (id[i]=='/') {
```
formatting

orgads replied:
Done

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,768 @@
+	it._id = id;
+	it._bmp = _icons[pic];
+	if (!it._bmp) {
+		it._bmp = Bitmap::create(pic + ".tga");
```
tga files?? what for?

aquadran commented on engines/grim/hotspot.h:
```diff
@@ -0,0 +1,161 @@
+	HotspotMan();
+	virtual ~HotspotMan();
+
+	enum ControlMode { Normal = 0,
```
formatting

aquadran commented on engines/grim/lua.cpp:
```diff
@@ -83,6 +83,14 @@ void LuaObjects::addNil() {
 	_objects.push_back(obj);
 }
 
+void LuaObjects::add(const float *ar, int len) {
```
changes to Lua will not be accepted. patches will needed to be modified to use unmodified Lua core

aquadran replied:
but, this is something for later in this PR, let's focus on other stuff first

orgads replied:
This is used in hotspot.cpp.

aquadran replied:
ah, never mind, I didn't look closely, I thought it was Lua core

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,768 @@
+	loadFlashBitmaps();
+	return true;
+}
+void HotspotMan::saveState(SaveGame *savedState) {
```
missing line space

orgads replied:
Done

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,768 @@
+		button = 2;
+	else if (ev.type == Common::EVENT_MBUTTONDOWN)
+		button = 3;
+	if (ev.kbd.hasFlags(Common::KBD_ALT))
```
why we add ATL and CTRL?

orgads replied:
Same big commit :/

aquadran replied:
I'll look into

aquadran replied:
can you tell me for what ALT and CTRL are used here? they are pushed into Lua some unknown reasons. since I don't see Lua script source, hard to guess

orgads replied:
This is the lua bytecode.
[mouse.zip](https://github.com/scummvm/scummvm/files/7464089/mouse.zip)

I can't decompile it. :/

aquadran replied:
you can use 'delua' from residualvm-tools, tools from scummvm-tools are not yet enabled.
I briefly looked into, and things are started to worry me. that patch can not have any original scripts code. it needs to be reviewed :/ but later

orgads replied:
Awesome!

This is the source file: [mouse.zip](https://github.com/scummvm/scummvm/files/7464151/mouse.zip)

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,768 @@
+}
+
+// get min distance s for x0 + s * x1, y0 + t * y1
+static double lineLineDist(
```
I think we have something in math already

orgads replied:
Where?

aquadran replied:
never mind

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -1703,4 +1764,5 @@ void GrimEngine::setSaveMetaData(const char *meta1, int meta2, const char *meta3
 	_saveMeta3 = meta3;
 }
 
+
```
no needed change

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -956,6 +982,10 @@ void GrimEngine::doFlip() {
 	}
 }
 
+inline float dist(const Common::Point &a, const Common::Point &b) {
```
is this needed?

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -841,6 +859,7 @@ void GrimEngine::updateDisplayScene() {
 		updateNormalMode();
 	} else if (_mode == DrawMode) {
 		updateDrawMode();
+		handleUserPaint();
```
why this change?

aquadran replied:
I think this might be needed to updating cursor, I'll look closer

aquadran replied:
yes, but you can just press 'I' for inventory

orgads replied:
Wrong conversation. I don't mind removing it. Remastered doesn't activate the inventory on Ctrl+LMB/MMB.

I think I'll leave middle button though, it can be convenient.

aquadran replied:
ok

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -389,15 +396,23 @@ Common::Error GrimEngine::run() {
 	} else {
 		g_driver = createRenderer(640, 480);
 	}
+	_system->showMouse(false);
```
check if we need to revert it on back to launcher

orgads commented:
I placed the whole-commit-reverts in a separate commit for now. Can you please verify that it doesn't revert "too much", then I'll squash it?

orgads commented:
Ok, I pushed the separate-revert to another branch in my fork, and squashed the commits here. This is the revert commit https://github.com/orgads/scummvm/commit/739b7c01cd431ab3f9ae92c917389f19a5d1e288.

aquadran commented on engines/grim/grim.cpp:
```diff
@@ -1025,8 +1047,29 @@ void GrimEngine::mainLoop() {
+				_hotspotManager->hover(_cursor->getPosition());
+			} else if (type == Common::EVENT_MBUTTONDOWN ||
+			           (type == Common::EVENT_LBUTTONDOWN && event.kbd.hasFlags(Common::KBD_CTRL))) {
+				Common::KeyState kbd(Common::KEYCODE_i);
```
looks like CTRL + LMB is mapped to 'I'. I think we should handle now with generic keys mapper.

orgads replied:
It opens the inventory.

aquadran replied:
keeping only LMB sounds better

orgads replied:
Done

aquadran replied:
hmm, actually I wonder if middle button can be mapped in generic way

orgads replied:
I can't find an example for that. If you have a way, let me know.

aquadran replied:
default engine mapping:
act = new Action(kStandardActionMiddleClick, _("Middle Click"));
act->addDefaultInputMapping("MOUSE_MIDDLE");
act->setMiddleClickEvent();
engineKeyMap->addAction(act);

Grim mapping:
act = new Action("INVT", _("Inventory"));
act->setKeyEvent(KeyState(KEYCODE_d, 'i'));
act->addDefaultInputMapping("JOY_Y");
engineKeyMap->addAction(act);

I don't remember how it works, but maybe try add:
act->setMiddleClickEvent();

or overing engines default

orgads replied:
I looked at it, but from what I see, `setMiddleClickEvent()` discards the previously set key.

aquadran replied:
two separated mappings doesn't looks good, so ok, keep as it is, but only for middle button

orgads replied:
Yes, that's already done. Thanks.

aquadran commented:
> Ok, I pushed the separate-revert to another branch in my fork, and squashed the commits here. This is the revert commit [orgads@739b7c0](https://github.com/orgads/scummvm/commit/739b7c01cd431ab3f9ae92c917389f19a5d1e288).

I think all related to android changed and scaler. so I think from code looks ok.
I assume it still work fine on desktop?

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,722 @@
+				hs._type >= 10 || hs._type == 3 || button == 0) {
+				continue;
+			}
+			if (gid<=0 || !_hotobject[gid]._active)
```
formatting

aquadran commented:
> > Ok, I pushed the separate-revert to another branch in my fork, and squashed the commits here. This is the revert commit [orgads@739b7c0](https://github.com/orgads/scummvm/commit/739b7c01cd431ab3f9ae92c917389f19a5d1e288).
> 
> I think all related to android changed and scaler. so I think from code looks ok. I assume it still work fine on desktop?

orgads commented:
> > Ok, I pushed the separate-revert to another branch in my fork, and squashed the commits here. This is the revert commit [orgads@739b7c0](https://github.com/orgads/scummvm/commit/739b7c01cd431ab3f9ae92c917389f19a5d1e288).
> 
> I think all related to android changed and scaler. so I think from code looks ok. I assume it still work fine on desktop?

It looks ok. I only tested the first few screens though.

orgads commented:
> one more thing, is mouse.lab file also reverted to previous version without android commit?

No. Should it be?

aquadran commented:
> > one more thing, is mouse.lab file also reverted to previous version without android commit?
> 
> No. Should it be?

never mind, I looked into wrong commit

orgads commented:
@aquadran did you have a chance to inspect the lua file? Anything else outstanding?

orgads commented:
Rebased and applied clang-tidy modernize-use-nullptr.

Should I add a setting for enabling/disabling mouse interaction? I think it boils down to [not] loading mouse.lab, am I right?

orgads commented:
Added a setting for disabling mouse (enabled by default).

aquadran commented:
mouse mod must be disabled by default. also save game compatibility must account account enabled and disable mouse support.

orgads commented:
Why is it needed in save game? If I started a game without a mouse, but now I do want to play with the mouse - shouldn't it just work if I enable it?

orgads commented:
Changed the default to disabled.

aquadran commented on engines/grim/cursor.cpp:
```diff
@@ -0,0 +1,80 @@
+ *
+ */
+
+#include "engines/grim/cursor.h"
```
can you run formatting whole file?

orgads replied:
Done

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,724 @@
+#include "common/array.h"
+#include "common/system.h"
+
+namespace Grim {
```
can you run formatting whole file?

orgads replied:
Done

aquadran commented on engines/grim/hotspot.cpp:
```diff
@@ -0,0 +1,727 @@
+
+void HotspotMan::drawActive() {
+	if (_flashHS) {
+		unsigned int curTime = g_system->getMillis();
```
use uint32

sev- commented:
I do not see sources or diffs for the `mouse.lab` file. What is inside?

orgads commented:
> I do not see sources or diffs for the `mouse.lab` file. What is inside?

See the file I attached here: https://github.com/scummvm/scummvm/pull/3406#discussion_r741481916

orgads commented:
> > I do not see sources or diffs for the `mouse.lab` file. What is inside?
> 
> See the file I attached here: [#3406 (comment)](https://github.com/scummvm/scummvm/pull/3406#discussion_r741481916)

Here, I pushed the source to my fork: https://github.com/orgads/scummvm/blob/grim-mouse-source/dists/engine-data/grim-mouse.lua

sev- commented:
Thanks! By looking at it, it seems to me that it is a decompiled and patched original source. Is this the correct assumption?

orgads commented:
If I understand correctly, it adds new functions that are used by the engine. For example [here](https://github.com/scummvm/scummvm/blob/3fed43de5968795d042c154f1f275b0e19e08114/engines/grim/hotspot.cpp#L333).

sev- commented:
> If I understand correctly, it adds new functions that are used by the engine. For example [here](https://github.com/scummvm/scummvm/blob/3fed43de5968795d042c154f1f275b0e19e08114/engines/grim/hotspot.cpp#L333).

That is fine that it adds new functions. But the fact that it contains the rest, e.g. we're redistributing a big chunk of the original source is not acceptable. We just cannot go with it. The very presence of this file in the fork is not violating the original copyright. Please, remove it.

The only way this could be accepted is if somebody makes sure that:
  1. No significant portion of the original scripts is present in the added lines (e.g. presence of scattered individual lines is OK, having whole chunks or lengthy functions of the original scripts is _not_ OK)
  2. A bindiff against the original is put into the repository, just as we do with grim-patch.lab (see `dists/engine-data/patches`)

aquadran commented:
this PR needs to be reworked for save game backward compatibility.

aquadran reviewed (changes requested):
this PR needs to be reworked for save game backward compatibility.

aquadran commented:
I reviewed Lua patches, they looks fine. but they heavily integrate, save game compatibility from old save to enabled mouse is practically not possible. patches also contain some delta not related to mouse support. possible they never changes from master. patches needs to be cleaned up

aquadran commented:
this could be implemented in future, but it needs to be done clean way with understanding what is changed and how it affect to game

orgads commented:
I wish you told me that before I spent so many hours on this patch. :(

This is a patch of all lua files (all are in DATA000.LAB): https://gist.github.com/orgads/ea684e04a44b4f30eb93821555ef2714

For some reason, `dlg_dom2.lua` seems to be broken. The binary patch is very small, but delua fails to parse it, and shows only part of the first line. I'm not sure if the patch really breaks the lua file, or is it a bug in delua.
